validator/src/main.rs

```diff
@@ -0,0 +1,249 @@
+use std::{net::SocketAddr, process::exit, sync::Arc, time};
+
+use anyhow::{anyhow, bail, ensure, Context, Result};
+use clap::Parser;
+use rand::distributions::{Alphanumeric, DistString};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream},
+    sync::Notify,
+};
+use tracing::{debug, error, info, Instrument};
+
+#[derive(Parser)]
+#[clap(version)]
+struct Args {
+    #[clap(
+        long,
+        env = "CNI_VALIDATOR_LOG_LEVEL",
+        default_value = "cni_validator=info,warn"
+    )]
+    log_level: kubert::LogFilter,
+
+    #[clap(long, env = "CNI_VALIDATOR_LOG_FORMAT", default_value = "plain")]
+    log_format: kubert::LogFormat,
+
+    #[clap(long, default_value = "0.0.0.0:4140")]
+    outbound_proxy_addr: SocketAddr,
+
+    #[clap(long)]
+    target_addr: SocketAddr,
+
+    #[clap(parse(try_from_str = parse_timeout), long, default_value = "120s")]
+    timeout: std::time::Duration,
+}
+
+// ERRNO 95: Operation not supported
+const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
+    let Args {
+        log_level,
+        log_format,
+        outbound_proxy_addr,
+        target_addr,
+        timeout,
+    } = Args::parse();
+
+    log_format.try_init(log_level)?;
+
+    let rng_resp = Alphanumeric.sample_string(&mut rand::thread_rng(), 8);
+    info!(%outbound_proxy_addr, %target_addr, "Validating outbound traffic is redirected to proxy's outbound port");
+
+    let (shutdown_tx, shutdown_rx) = kubert::shutdown::sigint_or_sigterm()?;
+    let notify_ready = Arc::new(Notify::new());
+
+    let server_task = tokio::spawn(outbound_serve(
+        outbound_proxy_addr,
+        notify_ready.clone(),
+        shutdown_rx,
+        rng_resp.clone(),
+    ));
+
+    let validation_task = tokio::spawn(validate_outbound_redirect(
+        target_addr,
+        notify_ready,
+        timeout,
+        rng_resp,
+    ));
```

Comment:
I don't think we really need to have synchronization (`notify_ready`) between these tasks. Instead, it seems simpler to bind the listening address before spawning the server so that we know that any connection attempts will hit that server.

Also, I don't think there's any need to spawn the client (validation) task. We can just await on the future in the select block directly...